dawn/docs/parallel_model.rst

```diff
@@ -0,0 +1,251 @@
+
+Formal Definition of the Dawn Parallel Model
+============================================
+
+Contract to the user from HIR
+-----------------------------
+
+Dawn will generate code based on the operations passed from high-level
+intermediate representation. It's specification can be found `here`_.
+
+Dependencies
+~~~~~~~~~~~~
+
+Generally, stencil computation in dawn differentiates between horizontal
+and the vertical dimension. The assumption is that individual statements
+of a stencil are embarrassingly parallel in the horizontal. For that
+reason, there is no guarantee that any horizontal point's computation
+will be finished before any other in each user statement.
+
+Vertical Dependencies
+^^^^^^^^^^^^^^^^^^^^^
+
+Each vertical region of a stencil specifies a loop order of execution.
+Dawn will uphold the following rules:
+
+-  If the loop order is forward, any statement with a vertical
+   dependency on other statements executed on a vertical level with
+   index ``k`` will be executed after each statement that it depends on
+   that appears previously in the current vertical region or in any
+   previous vertical region was executed on every vertical level with
+   indices ``k-1`` and smaller.
+-  If the loop order is backward, any statement with a vertical
+   dependency on another statement executed on a vertical level with
+   index ``k`` will be executed after each statement that it depends on
+   that appears previously in the current vertical region or in any
+   previous vertical region was executed on every vertical level with
+   indices ``k+1`` and bigger.
+
+An example of this would be the following
+
+.. code:: c++
+
+   1 vertical_region(k_start, k_end)
+   2   one = 1
+   3 vertical_region(k_start, k_end)
+   4   two = one[k - 1]
+
+Here, for every level with index ``k``, line two will be executed before
```

Comment:
Following the discussions above, what is ambiguous in the text I think is that both:
```
vertical_region(k_start, k_end)
   one = 1
vertical_region(k_start, k_end)
   two = one[k - 1]
```
and 
```
vertical_region(k_start, k_end)
   one = 1
   two = one[k - 1]
```
are compatible with the sentence `line two will be executed before`. Do we need to introduce the concept of a solver and add
`in case of a solver access, all writes into that field of previous vertical_regions are guaranteed to have been processed for the entire k domain before computing the solver.` ?

Reply:
They will do the same, but they are not the same (if you have `one[k + 1]`, they do different things in the formulation I am aiming for). But yes, I agree, that we should probably aim for a better formulation which explicitly specifies solvers; but first we might want to fix the current formulation (whether the full fix should be implemented or not - I don't know).